core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -368,19 +369,22 @@ public void closed(Status status) {
     }
 
     private void closedInternal(Status status) {
+      Throwable cancelCause = null;
       try {
         if (status.isOk()) {
           listener.onComplete();
         } else {
           call.cancelled = true;
           listener.onCancel();
+          // The status will not have a cause in all failure scenarios but we want to make sure
+          // we always cancel the context with one to keep the context cancelled state consistent.
+          cancelCause = status.getCause() != null ? status.getCause()
```

Comment:
I think we want to ignore the status completely, as we don't normally share that information with the application. So maybe:

```java
cancelCause = InternalStatus.asRuntimException(Status.CANCELLED.withDescription("RPC cancelled"), null, false);
```